sov-modules/sov-modules-impl/prover-incentives/src/call.rs

```diff
@@ -0,0 +1,170 @@
+use anyhow::Result;
+use bank::Coins;
+use borsh::{BorshDeserialize, BorshSerialize};
+use sov_modules_api::CallResponse;
+use sov_state::WorkingSet;
+use sovereign_sdk::zk::traits::Zkvm;
+use std::fmt::Debug;
+
+use crate::ProverIncentives;
+
+/// This enumeration represents the available call messages for interacting with the `ExampleModule` module.
+#[derive(BorshDeserialize, BorshSerialize, Debug, PartialEq)]
+// TODO: allow call messages to borrow data
+// https://github.com/Sovereign-Labs/sovereign/issues/274
+pub enum CallMessage {
+    BondProver(u64),
+    UnbondProver,
+    VerifyProof(Vec<u8>),
+}
+
+impl<C: sov_modules_api::Context, Vm: Zkvm> ProverIncentives<C, Vm> {
+    /// A helper function for the `bond_prover` call. Also used to bond provers
+    /// during genesis when no context is available.
+    pub(super) fn bond_prover_helper(
+        &self,
+        bond_amount: u64,
+        prover: &C::Address,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse> {
+        let mut response = CallResponse::default();
+
+        // Transfer the bond amount from the sender to the module's address.
+        // On failure, no state is changed
+        let coins = Coins {
+            token_address: self
+                .bonding_token_address
+                .get(working_set)
+                .expect("Bonding token address must be set"),
+            amount: bond_amount,
+        };
+        self.bank
+            .transfer_from(prover, &self.address, coins, working_set)?;
+
+        // Update our record of the total bonded amount for the sender.
+        // This update is infallible, so no value can be destroyed.
+        let old_balance = self
+            .bonded_provers
+            .get(prover, working_set)
+            .unwrap_or_default();
+        let total_balance = old_balance + bond_amount;
+        self.bonded_provers.set(prover, total_balance, working_set);
+
+        // Emit the bonding event
+        response.add_event(
+            "bonded_prover",
+            &format!("new_deposit: {bond_amount:?}. total_bond: {total_balance:?}"),
+        );
+
+        Ok(response)
+    }
+
+    /// Try to bond the requested amount of coins from context.sender()
+    pub(crate) fn bond_prover(
+        &self,
+        bond_amount: u64,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse> {
+        self.bond_prover_helper(bond_amount, context.sender(), working_set)
+    }
+
+    /// Try to unbond the requested amount of coins with context.sender() as the beneficiary.
+    pub(crate) fn unbond_prover(
```

Comment:
My spell checker does not like work "unbond". 

Maybe words "bind" and "unbind" would be clearer. 

But I am not native english speaker.

Reply:
Yeah, `bond`ing and `unbond`ing aren't standard English - but they're widely used to describe this kind of thing in blockchain: https://github.com/tendermint/tendermint/issues/2653